azure-iot-device/azure/iot/device/provisioning/internal/polling_machine.py

```diff
@@ -0,0 +1,325 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+import logging
+import uuid
+from threading import Timer
+from transitions import Machine
+from ..transport import constant
+
+from .action import SubscribeAction, PublishAction, UnsubscribeAction
+from .request_response_provider import RequestResponseMachine
+
+
+logger = logging.getLogger(__name__)
+
+
+class PollingMachine(object):
+    """
+    Class that is responsible for sending the initial registration request and polling the
+    registration process for constant updates.
+    """
+
+    def __init__(self, state_based_provider):
+        """
+        :param state_based_provider: The state machine based provider.
+        """
+        self._polling_timer = None
+        self._query_timer = None
```

Comment:
>self._query_timer = None [](start = 8, length = 24)

Is there only ever one timer that needs to run at a time?

Reply:
both must be run during polling , only 1 must be run during initial requesting

---
In reply to: [277823100](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277823100) [](ancestors = 277823100)

Reply:
No, I understand polling and query are different timers. I mean, under this design, there's only one query timer, and only one polling timer. Is that okay? What if something happens that overwrrites the timer, e.g. calling register() again.

---
In reply to: [277878077](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277878077) [](ancestors = 277878077,277823100)

Reply:
Still want an answer to this question.

---
In reply to: [277887263](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277887263) [](ancestors = 277887263,277878077,277823100)

Reply:
calling register again should throw an error. It is undefined yet what it will throw , based on what state the machine is.

The only way to actually call register again would ne instantiate a new client or cancel and then register

---
In reply to: [282608808](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r282608808) [](ancestors = 282608808,277887263,277878077,277823100)